### PR TITLE
CRM-20569: Fix membership on partially paid payment via backoffice form

### PR DESCRIPTION
**Pull request by monishdeb**
* [CRM-20569: Record Payment for partially paid membership: update amt, fix deferred ](https://issues.civicrm.org/jira/browse/CRM-20569)

### CODE REVIEW

**Comment by monishdeb** [comment]
The related UT is expected to fail as it dependent on #11000

File: CRM/Member/Form/Membership.php [hunk 1]
```diff
@@ -121,17 +121,14 @@ class CRM_Member_Form_Membership extends CRM_Member_Form {
    */
   public static function getSelectedMemberships($priceSet, $params) {
     $memTypeSelected = array();
-    $priceFieldIDS = self::getPriceFieldIDs($params, $priceSet);
-    if (isset($params['membership_type_id']) && !empty($params['membership_type_id'][1])) {
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: This line was to catch 'non price set' - we need to be sure that still works  (I know we have been working to disolve the distinction & perhaps we already have)

Comment (eileenmcnaughton) [reply CRM/Member/Form/Membership.php]: ah I see - handled in self::getPriceFieldIDs

**Comment by monishdeb** [comment]
Jenkins test this please

File: CRM/Member/BAO/Membership.php [hunk 2]
```diff
@@ -394,6 +394,9 @@ public static function create(&$params, &$ids, $skipRedirect = FALSE) {
         $params['line_item'],
         CRM_Utils_Array::value('contribution', $params)
       );
+      if (!empty($params['contribution'])) {
```
Comment (eileenmcnaughton) [inline CRM/Member/BAO/Membership.php]: I'm not sure we should be enhancing ways to add payments when creating a membership - this doesn't feel in sync with the general principal of creating a pending 'order' and then add payments

**Comment by eileenmcnaughton** [comment]
Ug - is this PR half way through an effort to add partial payments to memberships - I feel like I don't know when we started adding partial payments against memberships or what was agreed

**Comment by monishdeb** [comment]
@eileenmcnaughton no this PR is complete and after applying the patch it successfully records the partial payment and updated the related contribution and membership correctly. Earlier the partial payment is no longer extended for backoffice mebership whereas as per the patch, its all there, we just need to format the submitted values before passing to BAO fn and record the necessary the payment.

File: CRM/Member/BAO/Membership.php (hunk 2 above)
Comment (monishdeb) [reply CRM/Member/BAO/Membership.php]: We are already adding payments with status pending for other event/contribution backoffice. Likewise here we are just recording a pending payment. Isn't this correct?

**Comment by eileenmcnaughton** [comment]
Hmm - so this seems like a new feature? Extend partial payments to contribution?

**Comment by monishdeb** [comment]
Umm, extend partial payment for backoffice membership to be precise :)

**Comment by eileenmcnaughton** [comment]
OK - I need to take a step back on this & figure out where it is coming from because I saw in another PR we managed to get rid of the concept  of entering 'partial payment' as a status for contribution & the idea we are adding it into another  form without disambiguating between payment & contribution bothers me. If https://github.com/civicrm/civicrm-core/pull/10776 hadn't stalled we would have a clear path forward on this - I'm going to try to discuss with @colemanw later on

**Comment by monishdeb** [comment]
No problem, as per my understanding, for backoffice membership/event, payment means the recording the related contribution. And for backoffice contribution, we are not supporting partial payment as I can recall we have hidden 'Partially paid'  status.

**Comment by JoeMurray** [comment]
From my perspective, we currently need and support the idea of a contribution being partially paid. This has been in core for a long time for partially paid contributions related to event registrations, and we had approval from core team to extend partial payments to contributions and memberships. I would distinguish between partial payment status and partially paid status. For example, editting a contribution that has status of Completed to increase the total amount should result in a status of partially paid. 

There is a promise from Jamie to totally rethink contribution status field, but that is more of a lexim thing. 

Can you be more precise about what part of the contribution status value list and supported status transitions you'd like to change? I note that the spec for this is at https://wiki.civicrm.org/confluence/pages/viewpage.action?pageId=78151699 and you were actually the person who updated the page in 2016 to document the usage of Partially Paid status after WA changes added this.

**Comment by eileenmcnaughton** [comment]
I've been thinking about this & I feel we stalled on Jamie's input on moving the payment block to something more appropriate on the contribution form - but when I look back at that PR it would be fairly simple to move it to an extension - since we did so much work simplifying core to facilitate it the final changes are actually pretty small & I've been thinking to go back to that & look again. I don't think we can improve the ability to make multiple payments against any entity without having a way to represent the many-to-one relationship in the form properly, and I think resolving that in the contribution form first makes sense

**Comment by eileenmcnaughton** [comment]
@monishdeb I'm going to close this because a) it's stale & b) I think we should sit on this for a release or 2 which the payment form change beds in & then do this in that context

**Pull request closed without merging**